File: socos/services/zeit.py
"""
Rechnen mit Zeiten.

**Gerundet wird erst hier, nie beim Speichern.** Gespeichert werden Start und
Ende sekundengenau; wer beim Erfassen rundet, kann die Rundung nicht mehr
zurücknehmen, wenn die Regel sich ändert.

Dieselbe Rundung gibt es ein zweites Mal im Frontend
(`frontend/src/basis/zeit.ts`), für die Anzeige. Beide sind getestet und müssen
dasselbe Ergebnis liefern.
"""

from datetime import date, datetime, time, timedelta
from decimal import ROUND_HALF_UP, Decimal

from django.utils import timezone

RUNDUNG_MINUTEN = 5


def auf_fuenf_minuten(sekunden):
    """
    Auf volle 5 Minuten, kaufmännisch. Genau 2,5 Minuten Rest gehen auf.

    Warum aufrunden statt abschneiden: Sonst verschwände über eine Woche
    verteilt eine Viertelstunde, die jemand tatsächlich gearbeitet hat.
    """
    schritt = RUNDUNG_MINUTEN * 60
    if sekunden <= 0:
        return 0
    return int(Decimal(sekunden / schritt).quantize(Decimal("1"), rounding=ROUND_HALF_UP)) * schritt


def summe_gerundet(sekundenliste):
    """
    Erst summieren, dann runden.

    Andersherum wären drei Buchungen zu je zwei Minuten zusammen null — pro
    Woche eine halbe Stunde Arbeit, die still verschwindet.
    """
    return auf_fuenf_minuten(sum(sekundenliste))


def als_stunden(sekunden):
    """Dezimalstunden für den Zeitnachweis, als Decimal — nie als float."""
    return (Decimal(auf_fuenf_minuten(sekunden)) / Decimal(3600)).quantize(
        Decimal("0.01"), rounding=ROUND_HALF_UP
    )


def als_stunden_text(sekunden):
    """
    Dezimalstunden als deutscher Text: 26,25 — mit Komma.

    Der Punkt aus `str(Decimal)` ist in einem deutschen Dokument schlicht
    falsch und wird beim Übertragen in eine Tabelle zur Fehlerquelle.
    """
    return f"{als_stunden(sekunden)}".replace(".", ",")


def als_dauer(sekunden):
    """Sekunden als 3:25 (Stunden:Minuten)."""
    minuten = max(0, int(sekunden)) // 60
    return f"{minuten // 60}:{minuten % 60:02d}"


def dauer(buchung, jetzt=None):
    """
    Dauer einer Buchung in Sekunden. Eine laufende zählt bis `jetzt`.

    Ein Entwurf zählt **null**: Er ist nicht bestätigt, und eine unbestätigte
    Buchung darf keine Auswertung beeinflussen.
    """
    if buchung.ist_entwurf:
        return 0
    bis = buchung.ende or (jetzt or timezone.now())
    return max(0, int((bis - buchung.start).total_seconds()))


def tagesende(zeitpunkt):
    """
    Der letzte Moment des Tages, an dem `zeitpunkt` liegt — in lokaler Zeit.

    Nicht `start + 24 h`: An den beiden Umstellungstagen hat ein Tag 23 oder
    25 Stunden, und eine Buchung landete sonst eine Stunde daneben.
    """
    lokal = timezone.localtime(zeitpunkt)
    naechster = timezone.make_aware(
        datetime.combine(lokal.date() + timedelta(days=1), time.min),
        timezone.get_current_timezone(),
    )
    return naechster - timedelta(seconds=1)


def woche_um(tag=None):
    """Montag bis Sonntag der Woche, in der `tag` liegt."""
    tag = tag or timezone.localdate()
    montag = tag - timedelta(days=tag.weekday())
    return montag, montag + timedelta(days=6)


def monatsgrenzen(monat: date):
    """Erster und letzter Tag des Monats, in dem `monat` liegt."""
    erster = monat.replace(day=1)
    if erster.month == 12:
        naechster = erster.replace(year=erster.year + 1, month=1)
    else:
        naechster = erster.replace(month=erster.month + 1)
    return erster, naechster - timedelta(days=1)

File: frontend/src/basis/zeit.ts
/**
 * Rechnen mit Zeiten — die Anzeige-Seite.
 *
 * **Gerundet wird erst in der Auswertung, nie beim Speichern.** Gespeichert
 * werden Start und Ende sekundengenau; wer beim Erfassen rundet, kann die
 * Rundung nicht mehr zurücknehmen, wenn die Regel sich ändert.
 *
 * Die maßgebliche Rundung für Berichte steht serverseitig in
 * socos/services/zeit.py. Diese hier ist für die Anzeige und muss dasselbe
 * Ergebnis liefern — beide sind getestet.
 */

export const RUNDUNG_MINUTEN = 5;

/** Sekunden als laufende Uhr: 01:23:45 */
export function alsUhr(sekunden: number): string {
  const s = Math.max(0, Math.floor(sekunden));
  const zwei = (n: number) => String(n).padStart(2, "0");
  return `${zwei(Math.floor(s / 3600))}:${zwei(Math.floor(s / 60) % 60)}:${zwei(s % 60)}`;
}

/** Sekunden als Dauer: 3:25 (Stunden:Minuten) */
export function alsDauer(sekunden: number): string {
  const minuten = Math.floor(Math.max(0, sekunden) / 60);
  return `${Math.floor(minuten / 60)}:${String(minuten % 60).padStart(2, "0")}`;
}

/**
 * Auf volle 5 Minuten, kaufmännisch. Genau 2,5 Minuten Rest gehen auf — sonst
 * verschwände in einer Woche eine Viertelstunde, die jemand gearbeitet hat.
 */
export function aufFuenfMinuten(sekunden: number): number {
  const schritt = RUNDUNG_MINUTEN * 60;
  return Math.round(Math.max(0, sekunden) / schritt) * schritt;
}

/** Erst summieren, dann runden. Andersherum summieren sich die Rundungsfehler. */
export function summeGerundet(einzelne: number[]): number {
  return aufFuenfMinuten(einzelne.reduce((a, b) => a + b, 0));
}

/** Dezimalstunden für den Zeitnachweis: 3,25 h */
export function alsStunden(sekunden: number): string {
  return (aufFuenfMinuten(sekunden) / 3600).toFixed(2).replace(".", ",");
}

/**
 * Heute als `2026-09-08` für ein `<input type="date">`.
 *
 * Nicht `toISOString().slice(0, 10)`: Das rechnet nach UTC, und westlich von
 * Greenwich — oder hier abends nach 22 Uhr Sommerzeit — steht dann der
 * falsche Tag im Feld. Gebraucht wird der Tag der Uhr an der Wand.
 */
export function heuteAlsDatum(jetzt = new Date()): string {
  const zwei = (n: number) => String(n).padStart(2, "0");
  return `${jetzt.getFullYear()}-${zwei(jetzt.getMonth() + 1)}-${zwei(jetzt.getDate())}`;
}

/** Der Tag eines Zeitpunkts als `2026-09-15` — für ein `<input type="date">`. */
export function tagAusZeitpunkt(iso: string): string {
  return heuteAlsDatum(new Date(iso));
}

/** Die Uhrzeit eines Zeitpunkts als `09:02` — für ein `<input type="time">`. */
export function uhrzeitAusZeitpunkt(iso: string): string {
  const d = new Date(iso);
  const zwei = (n: number) => String(n).padStart(2, "0");
  return `${zwei(d.getHours())}:${zwei(d.getMinutes())}`;
}

/**
 * Tag und Uhrzeit zu einem Zeitpunkt — aus den Bestandteilen, nicht über
 * `new Date("2026-09-15T09:02")`.
 *
 * Der Zeichenkettenweg ist in alten Safari-Fassungen als UTC gelesen worden;
 * daraus wird im Sommer eine Buchung zwei Stunden daneben, und das fällt an
 * einer einzelnen Zeile niemandem auf.
 */
export function alsZeitpunkt(tag: string, uhrzeit: string): Date {
  const [jahr, monat, tagnr] = tag.split("-").map(Number);
  const [stunde, minute] = uhrzeit.split(":").map(Number);
  return new Date(jahr, monat - 1, tagnr, stunde, minute, 0, 0);
}

export type Spanne = { start: Date; ende: Date; ueberNacht: boolean };

/**
 * Beginn und Ende aus einem Tag und zwei Uhrzeiten.
 *
 * **Ein Ende, das nicht nach dem Beginn liegt, meint den Folgetag** — 22:00
 * bis 01:00 sind drei Stunden, keine negativen einundzwanzig. Die Oberfläche
 * schreibt den Folgetag ausdrücklich hin; stillschweigend wäre es genau die
 * Sorte Zahl, die plausibel aussieht und um einen Tag danebenliegt.
 *
 * Fehlt etwas, kommt `null` — der Aufrufer sagt dann, was fehlt.
 */
export function spanne(tag: string, von: string, bis: string): Spanne | null {
  if (!tag || !von || !bis) return null;
  const start = alsZeitpunkt(tag, von);
  let ende = alsZeitpunkt(tag, bis);
  const ueberNacht = ende <= start;
  if (ueberNacht) ende = new Date(ende.getTime() + 24 * 60 * 60_000);
  return { start, ende, ueberNacht };
}

/**
 * Eine Uhrzeit plus Minuten, als `HH:MM`. Über Mitternacht hinaus fängt sie
 * wieder bei 00:00 an — welcher Tag gemeint ist, entscheidet `spanne`.
 */
export function plusMinuten(uhrzeit: string, minuten: number): string {
  const [stunde, minute] = uhrzeit.split(":").map(Number);
  const gesamt = (((stunde * 60 + minute + minuten) % 1440) + 1440) % 1440;
  const zwei = (n: number) => String(n).padStart(2, "0");
  return `${zwei(Math.floor(gesamt / 60))}:${zwei(gesamt % 60)}`;
}
